sql-serializer/src/main/java/io/cdap/pipeline/sql/serializer/bigquery/BigQuerySQLSerializer.java

```diff
@@ -0,0 +1,247 @@
+
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Serializes a {@link StructuredQuery} into a {@link String} able to be executed upon Google BigQuery.
+ */
+public class BigQuerySQLSerializer implements SQLSerializer {
+  public String getSQL(StructuredQuery query) {
+    if (query == null) {
+      throw new IllegalArgumentException("Query cannot be null");
+    }
+    StringBuilder b = new StringBuilder();
+    return appendQuerySQL(b, query, false).toString();
+  }
+
+  public String getSQL(StructuredQuery query, Table temporaryTable) {
+    if (query == null) {
+      throw new IllegalArgumentException("Query cannot be null");
+    }
+    if (temporaryTable.hasAlias()) {
+      throw new IllegalArgumentException("Cannot have alias on temporary table");
+    }
+    if (temporaryTable.hasProject() || temporaryTable.hasDatabase()) {
+      throw new IllegalArgumentException("Temporary tables cannot be fully qualified");
+    }
+    StringBuilder b = new StringBuilder();
+    b.append("CREATE TEMPORARY TABLE ");
+    appendTableSQL(b, temporaryTable, false);
+    b.append(" AS ");
+    appendQuerySQL(b, query, false);
+    return b.toString();
+  }
+
+  private StringBuilder appendQuerySQL(StringBuilder b, StructuredQuery query, boolean isSubquery) {
+    // Keep track of which columns were defined in the current nesting level since they may also be used in filter
+    HashSet<Column> definedColumns = new HashSet<>();
+    if (isSubquery) {
+      b.append('(');
+    } else if (query.hasAlias()) {
+      throw new IllegalArgumentException("Cannot have an alias on the main outer query");
+    }
+    b.append("SELECT ");
+    for (int i = 0; i < query.getColumns().size(); i++) {
+      Column col = query.getColumns().get(i);
+      if (col.hasAlias()) {
+        definedColumns.add(col);
+      }
+      appendColumnSQL(b, col, true, false);
+      if (i < query.getColumns().size() - 1) {
+        b.append(", ");
+      }
+    }
+    b.append(" FROM ");
+    appendQueryableSQL(b, query.getQueryFrom());
+    if (query.getFilter() != null) {
+      b.append(" WHERE ");
+      appendFilterSQL(b, query.getFilter(), definedColumns);
+    }
+    if (isSubquery) {
+      b.append(')');
+      if (query.hasAlias()) {
+        b.append(" AS ");
+        b.append(query.getAlias());
+      }
+    } else {
+      b.append(';');
+    }
+    return b;
+  }
+
+  private StringBuilder appendColumnSQL(StringBuilder b, Column c,
+                                        boolean defineAliasIfExists, boolean useAliasIfExists) {
+    if (defineAliasIfExists && useAliasIfExists) {
+      throw new IllegalArgumentException("Unable to use and define an alias at the same time");
+    }
+    if (c.hasSource() && c.getSource().hasAlias()) {
+      b.append(c.getSource().getAlias());
+      b.append('.');
+    }
+    if (c.hasAlias() && useAliasIfExists) {
```

Comment:
I don't understand this logic, when would you want to do each of these conditions? Can you add comments with examples?

Reply:
I removed all of this from appendColumn, but will add an explanation for appendTable.